### DISCUSSION TITLE
Use local DNS Record (Cloudflare Tunnel 100MB Limit Workaround)

### DISCUSSION THREAD

**Discussion by Yavieh** [immich-app/immich#13175] [General]
In this discussion, we’ll talk about using local DNS records to bypass this limit within the local network. By setting up a local DNS, we can route larger traffic locally while still using Cloudflare Tunnel for external connections.

Feel free to share your experiences and ideas!

_This discussion was moved from here https://github.com/immich-app/immich/discussions/1674_

**Comment by Yavieh** [comment]
```
I've just set up 2nd option minutes ago and it is working great...when i am outside it uses cloudflare tunnel (only uploads <100mb) and when at home it goes locally

I am not an expert but here we go:

Go to your pihole: Local DNS - DNS Records and add:
-Domain: yourdomainorsubdomain.com
-Ip adress: localipwhereimmichisinstalled (without port)...Add it.
-Optional: flush dns at PC (open cmd console and write: ipconfig /flushdns) and at pihole update gravity under tools menu and also under settings menu go to system flush network table and restar dns resolver.

Install Nginx proxy manager and add a proxyhost:
-Domain name: yourdomainorsubdomain.com
-scheme: http
Forward Hostname / IP: localipwhereimmichisinstalled
-Forward port: portofimmich (usually 2283)
Tick: block common exploits and websockets support and Save.
No custom locations, no SSL, no Advanced settings is needed.

You can test it sending a ping at your local network to your domain and it should give you local ip. Use a vpn or mobile data and it should give you cloudflare's ip

I think that's all.
```

This was the last comment from `camafive` which i still got some Questions on. 
How does nginx come in play here? The locas DNS Record directly links to the immich ip, so the nginx will never be reached or am i missing here something?

**Reply by camafive** [reply]
> ```
> I've just set up 2nd option minutes ago and it is working great...when i am outside it uses cloudflare tunnel (only uploads <100mb) and when at home it goes locally
> 
> I am not an expert but here we go:
> 
> Go to your pihole: Local DNS - DNS Records and add:
> -Domain: yourdomainorsubdomain.com
> -Ip adress: localipwhereimmichisinstalled (without port)...Add it.
> -Optional: flush dns at PC (open cmd console and write: ipconfig /flushdns) and at pihole update gravity under tools menu and also under settings menu go to system flush network table and restar dns resolver.
> 
> Install Nginx proxy manager and add a proxyhost:
> -Domain name: yourdomainorsubdomain.com
> -scheme: http
> Forward Hostname / IP: localipwhereimmichisinstalled
> -Forward port: portofimmich (usually 2283)
> Tick: block common exploits and websockets support and Save.
> No custom locations, no SSL, no Advanced settings is needed.
> 
> You can test it sending a ping at your local network to your domain and it should give you local ip. Use a vpn or mobile data and it should give you cloudflare's ip
> 
> I think that's all.
> ```
> 
> This was the last comment from `camafive` which i still got some Questions on. How does nginx come in play here? The locas DNS Record directly links to the immich ip, so the nginx will never be reached or am i missing here something?

Hi, the Nginx proxy manager is to route the traffic to the correct app...because for example if your local ip is: 192.168.1.1 and you put only that at the local dns (adguard or pihole) the traffic dont know where to go because you probably have other apps under the same ip but on different ports, examples:

192.168.1.1:9000 app1
192.168.1.1:7463 app2

Where is traffic going to go if you only put the local ip if you have other apps under the same ip? So there is the need of Nginx proxy manager...it tells the traffic that comes from domain.com that the correct app is for example app2 (192.168.1.1:7463)

I hope you understood...sorry for my english (not native)

Btw: i said before that there is no need of ssl at nginx...well i had to set ssl because i was having problems to upload outside my network. Now after i set ssl it seems working as expected

**Reply by Yavieh** [reply]
So the nginx runs at the same ip as the Immich server but the nginx does on port 80?
For m my setup, the nginx is on a different ip, so in the dns route I will setup the dns record to the ip of the nginx right ?

**Reply by camafive** [reply]
I am not an expert so i dont know if that would work. But yes, nginx is at the same local machine on port 81

**Reply by Yavieh** [reply]
Okay thanks for the help.
What still confuses me is, the dns record ist only redirecting all traffic from 1 domain to the other. So from mydomain.com to localip. But the port would be 80 if I just call mydomain.com so it won’t be reached to nginx when that is running on 81 or not ?

**Reply by camafive** [reply]
The A record is to tell your devices inside your network...hey domain.com is hoted here (local ip) and then nginx proxy says...yes here i have that service but at this port (port of the service/app). Port 81 is only the Port where you reach npm, it is just like other apps that have their ports. Think about it like this:

You ask me: hey where does Andrew (service /app)live? And i say: he lives in that building over there (i am A record)...you got to the building and a receptionist tells you (nginx) that he indeed live there but at 2nd floor

**Reply by camafive** [reply]
Dont worry about ports, you only need npm installed locally with the other apps and thats it. It does not matter if nginx is at port 81 or whatever and the service is at port 500, 80 or whatever

**Reply by Yavieh** [reply]
Okay, but my Brain still has problems understanding exactly how it’s working. For me it does not makes sense how the npm is involved into the whole thing when it’s not talked to via port.

For me it only would make sense this way.
I got a domain mydomain.com
Npm is on ip 192.168.12.2 port 80
And Immich at ip 192.168.12.3 port 1234 or whatever port

then in my local dns I will say mydomain.com go to the npm (192.168.12.2) 
Then npm says I got an entry for that domain. Go here 192.168.12.3:1234

when the dns would directly retry the Immich ip for the domain so .3 it would never reach npm in th first place.

but yeah maybe I just don’t get it.

Thanks for your patience and help

**Reply by camafive** [reply]
It is just like that. You tell in the A record your local ip where all your apps/services are installed and npm just route the traffic to the correct port. Thats it. Have you tried it?

**Reply by Yavieh** [reply]
I will try it tomorrow when I’m back home.
Maybe I can get it working :) 
Thanks

**Reply by camafive** [reply]
Try it that way, it should work. If it doesnt let me know, we can do it via teamviewer or some if you want

**Reply by Yavieh** [reply]
Okay I will, thanks for your help so far.

**Reply by tazire** [reply]
I don't use pi hole but I can set local DNS on my udm pro. Should this work the same way?

**Reply by camafive** [reply]
I dont know what is udm pro but It should work. Give it a try and let us know

**Reply by tazire** [reply]
Unifi dream machine pro. It doesn't seem to work. I have my immich instance as a docker on an unraid server. I may need to give it more time after setting the local DNS for it all to take control. I have been using NPM with cloud flare proxy(not tunnel) and having the upload issues too. The local DNS works as it should (ping from inside the network sends it to the local IP)but NPM isn't picking up the request and sending it to the correct port. I had to add it as an host A record in DNS because unifi doesn't support cname additions just yet. I wonder if this might be a problem.

**Reply by camafive** [reply]
That is what you need...an A record...not a cname

**Reply by tazire** [reply]
Ok well then it's not working at present. I've set up the a name record and NPM works with external requests but it doesn't seem to be working with requests when local DNS is setup. I just get connection refused all the time.

**Reply by tazire** [reply]
But when I ping immich.domain.com it pings my server just fine. So the issue seems to be with npm for some reason

**Reply by camafive** [reply]
My set up is like this:
For local querys: A record at pihole telling that domain.com is hosted at for example: 192.168.1.155 and at NPM i have a proxy host that re-route traffic to the correct port of the app/service.

For external querys: cloudflare tunnel

Does your setup looks like that?

**Reply by tazire** [reply]
Yes. I had it setup perfectly when I wasn't trying to use local DNS. And connection works fine. I don't use a cloud flare tunnel but I was using cloud flare as a proxy. But I'm guessing that shouldn't be the issue anymore because the local DNS should give me a direct connection to my server... I have a feeling it's got to do with the internal port for npm not being 80 and 443. The internal ports for npm on my server is 1880 and 18443. 80 port forwarded to 1880 and 443 port forwarded to 18443. My port forwarding rules only apply to internet traffic I think. And that's where the problem is I believe. I think I'll have to use 80 and 443 on the npm container.

**Reply by camafive** [reply]
Try it, i am not an expert but i dont see what is the problem with the port of npm...if it is under the same local network of the other apps it should work.

**Reply by tazire** [reply]
I was correct. So essentially what was happening in my case my port forwarding rule only applies to internet traffic. When the local DNS request was coming in this rule didn't apply and NPM internally is listening to port 1880 and 18443. I changed these to 80 and 443 and changed my port forwarding etc and it now works

**Reply by Yavieh** [reply]
I also got it working for my setup 🎉
Currently I got cloudflare tunnel where I also define the routes to the local instances like Immich and next cloud and co.
Then I got a npm installed just for my local routing. The npm also has the ssl cert for my domain otherwise it would not work for me as in the app I set up the url via https.
Then I got a dns record in my local adguard which is redirecting my Immich domain to my local npm which then proxies that to the Immich server.
Works perfectly 🤩
Now i can upload files bigger then 100mb in my local network. 🎉

**Reply by tehniemer** [reply]
Can anyone help me get this set up in a pfSense router using its DNS resolver and HAProxy?

**Reply by Yavieh** [reply]
> Can anyone help me get this set up in a pfSense router using its DNS resolver and HAProxy?

Hey, im not familiar with that brand but i found [this](https://docs.netgate.com/pfsense/en/latest/services/dns/resolver-host-overrides.html) documentation which should be what you are searching for.

Im also not familiar with HAProxy. But it should work like any other reverse proxy i guess.

**Reply by tehniemer** [reply]
Thanks, I seem to have figured it out. I'll create a separate comment to outline the instructions for pfSense

**Comment by tehniemer** [comment]
I've managed to get this working using pfSense and its available packages. Here's how I went about it.

1. Install the `acme` and `haproxy` packages
2. Create an IP Alias to the Localhost interface, I used 192.168.100.1 for example. `Firewall > Virtual IPs`
3. Create a Host Override for the address pointing to whatever virtual IP you created in step 2. `Services > DNS Resolver`
4. Create an account key with your preferred ACME server. `Services > Acme > Account Key`
5. Create a certificate for your host/domain. I used a wildcard cert (*.domain.com), which forced the method to be a DNS challenge. But you could likely create a cert specific to the host without having to use DNS challenge. `Services > Acme > Certificates`
6. Click the "Issue/Renew" button when you are back on the certificates page
7. In HAProxy, create a backend with the address and port of your immich instance, leave the SSL boxes unchecked. `Services > HAProxy > Backend`
8. Create a frontend that listens on the IP from step 2 on ports 80 and 443. Check the SSL Offloading box in the entry for port 443. `Services > HAProxy > Frontend`
9. In the same frontend, create an access control list entry with the "host matches:" expression with a value of whatever the host and domain you used in the host override in step 3. Then in actions, set "Use Backend" with the condition set to the name of the ACL you just created and set the backend to the name of the backend you created in step 7. Finally, select the certificate you created in step 5 in the SSL Offloading section.

If all goes as planned, your service will be available on the domain of your choosing and when you ping it, it will return the IP Alias you set up instead of your public IP or IP of your registrar.

**Comment by SircasticFox** [comment]
I was fighting this issue for a while as well. Being behind a CGNAT and not having a public accessible IPv4 address I used cloudflare for immich to be reachable in countries where IPv6 coverage is quite bad (e.g. on vacation to backup my photos). But well, there comes to upload limit. Long story short, for this usecase unfortunately a local DNS doesn't really help me.

I solved this issue by completely getting rid of cloudflare and routing all the traffic through a VPS (that I already run anyway). My immich FQDN now points to my VPS IPv4 address which just routes the traffic to my Immich Instance. On the VPS I use Traefik for this.

Not a solution to keep cloudflare, but for people behind a CGNAT who used cloudflare to make Immich accessible everywhere.

**Comment by Neo11Neo** [comment]
I just found out that in the immich app under `settings > network` you are able to setup automatic url switching. This means you can directly point to your immich instance when you are connected to your local wifi.

**Reply by d1egoaz** [reply]
Thanks, this solved the issue for me, I guess I can live without uploading my videos over mobile.

**Comment by dsquare86** [comment]
I am using NPM for reverse proxy, Pihole w/ Unbound for DNS, and using DNS proxy from Cloudflare.

Solution was to add custom DNS Masq .conf file calling any address in my domain as local with a local DNS entry in Pihole pointing to NPM.

**Comment by jondycz** [comment]
My setup allows me to access immich over the same domain over SSL:
• locally - no limit
• from the internet (100MB limit)
• using Cloudflare One (zero trust warp) - no limit

Prerequisite: Adguard home to route your domains to local IPs
1) I installed nginxproxymanager and created an immich host - this allows local access with SSL
2) I installed cloudflare tunnel and pointed my domain to nginxproxymanager so that I can manage all my subdomain in one place - this allows for external access with 100MB limit
3) I connected my devices to cloudflare zero trust over the Cloudflare One app
4) IMPORTANT! - I created a firewall rule in the zero trust dashboard to forward my immich domain to the LOCAL IP of my nginxproxymanager. By default cloudflare One (warp) does not allow setting a custom DNS resolver (only for a business plan) so I couldn't use my local DNS server (Adguard) to resolve the immich IP to a local one. So devices connected through warp always tried to access immich externally through the cloudflare proxy. But using the firewall rule, all devices connected through warp now resolve the domain to a local IP that is not affected by the 100MB upload limit.

TLDR: When you need to upload files larger than 100MB, turn on Cloudflare One on your device and make sure to set up firewall rules so that your domain gets resolved to your local IP address and skips the cloudflare proxy

**Comment by Luggi1900** [comment]
## Problem Description

I tried to copy your setup, but it does not work for me.

## What I have done so far

1. **Set up TrueNAS SCALE**

2. **Installed the Immich app on TrueNAS**
   - TrueNAS app exposure ports: `8443` / `8080`

3. **Configured external access via Cloudflare**
   - Bought the domain `my.immich.com`
   - Created a Cloudflare Tunnel to access Immich from the internet

   This works fine, but I hit Cloudflare’s **100 MB upload limit**.

4. **Tried to bypass the limit using local DNS (Pi-hole)**
   - Installed the Pi-hole app on TrueNAS
   - Created a local DNS record:
     ```
     my.immich.com -> local.ip
     ```
   - Configured my FRITZ!Box to use `local.ip` as a DNS server

5. **Set up Nginx Proxy Manager for local HTTPS/HTTP**
   - Installed the Nginx Proxy Manager app on TrueNAS
   - Changed NPM’s ports to `443` / `80`
   - Added a Proxy Host with:
     - **Domain Name:** `my.immich.com`
     - **Scheme:** `http`
     - **Forward Hostname / IP:** `local.ip`
     - **Forward Port:** `30041` (Immich port on TrueNAS)
   - Tested both without SSL and with SSL
   - Added this config to *Advanced Settings*:
     - https://docs.immich.app/administration/reverse-proxy

## Current Issue

Even from inside the local network, accessing `my.immich.com` always goes through the **Cloudflare Tunnel** instead of using the local DNS override and the reverse proxy.

## DNS Debugging

When I run `nslookup my.immich.com`, I get the following result:

```text
Server:   UnKnown
Address:  2a00:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx

Name:     my.immich.com
Addresses:
          2a06:xxxx:xxxx::x
          2a06:xxxx:xxxx::x
          192.168.x.x
```

If i run `nslookup my.immich.com 192.168.x.x`, I get this result:

```text
Server:   my.immich.com
Address:  192.168.x.x

Name:     my.immich.com
Addresses:
          2a06:xxxx:xxxx::x
          2a06:xxxx:xxxx::x
          192.168.x.x
```

I tried disabling IPv6 on my FRITZ!Box and also blocking the AAAA record in Pi-hole.
In both cases, `my.immich.com` becomes completely unreachable from the local network instead of falling back to IPv4.

I would be very grateful if someone could help me solve this issue.

**Reply by Yavieh** [reply]
I would change your pi hole dns settings.
Set the immich dns domain to the ngnix proxy manager ip. The ngnix should then reroute your direktly to local immich ip.

To test if the setup works, create a second entry which with a subdomain which does not exist in cloudflare and only in your ngnix.

So for example
- my.immich.com exists both on cloudflare and your ngnix 
- my2.immich.com only in your nginx

In pihole you can either route every subdomain to the nginx or you create an entry for each subdomain.

So the flow should be like this:
Your device (pc, smartphone) -> fritz.box -> pihole dns -> nginx -> immich

When its still not working, let me know how you configured your dns settings in fritzbox.
You could also test to manually set the dns to pihole on your device network settings.

Let me know if it worked. Otherwise we need to take a deeper look.

**Reply by Yavieh** [reply]
I was reading your setup again and i guess the first part is already given, so you can skip that if true 😄

**Reply by jondycz** [reply]
You have to create a firewall rule if you want clooudflare tunnel to use your local DNS. Check my comment that describes how to do it in this thread

**Reply by Yavieh** [reply]
That should not be nessecarry if his local dns already routes the outgoing request to the local nginx.
I also dont have any cloudflare firewall setting.

**Reply by Luggi1900** [reply]
I made the following changes:

- I updated the Cloudflare Tunnel so it no longer points directly to Immich but instead to Nginx.
- I disabled IPv6 in my local network.
- I created a second entry using a subdomain that does not exist in Cloudflare and is only configured in my Nginx setup.
- I added a local DNS record in pi-hole for this domain
- I did not configure SSL for this domain (for `my.immich.com`, SSL is set up via Cloudflare using a DNS challenge).

Now, this is what happens when I try to access `my2.immich.com` in the browser:

- **Normal browser tab:** `ERR_SSL_UNRECOGNIZED_NAME_ALERT`
- **Private/incognito tab:** "`my2.immich.com` does not support a secure connection with HTTPS", with an option to continue. When I continue, I am redirected to Immich.

If I use a custom SSL certificate, it works fine in both cases.

When I try to access `my.immich.com`, this happens:

- **From within the LAN:** 
`ERR_SSL_UNRECOGNIZED_NAME_ALERT` or
`ERR_QUIC_PROTOCOL_ERROR`, changing every few refreshs
- **From the internet:** the default Nginx landing page ("Congratulations! ... you’re trying to access a host that isn’t set up yet")

**Reply by Yavieh** [reply]
My suggestions would be to configure ssl certs within your nginx, so then you have ssl connections for both outside and local.

The second thing is, im not sure if your dns is working as intended.
Pls try this to see if your dns reroute works. I just did the same with my adguard dns:

Take any domain for example test.immich.com and reroute that to google.com
When accessing thst domain now, it should show a google error page. If that works, your dns settings are working as expected. If non, pls set your pihole as the one and only dns in your device settings and test again.

Let me know the result.

**Reply by Luggi1900** [reply]
I actually managed to solve the issue a while ago and realized I forgot to post an update.

Just for info: I already had the SSL certs set up in Nginx via Cloudflare DNS challenge.

I honestly have no clue why the problem occurred in the first place because I don't really have the technical understanding for it, but I fixed it with the following steps:

I kept everything as described in my first post, except for the advanced settings – those aren't needed at all. The actual solution was the configuration in Pi-hole.

Instead of a single local DNS entry pointing `my.immich.com` to my `local.ip`, I did this:

1. Created a **CNAME record**: `my.immich.com` pointing to `immich.com`.
2. Created a **local DNS record**: `immich.com` pointing to `local.ip`.

With this setup, it finally works and the traffic stays local. Thanks for the help anyway :D
